File: internal/observability/gauges.go
package observability

import (
	"context"
	"fmt"
	"sync/atomic"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// PartitionGauge is one row in the per-partition gauge sample. The
// GaugeSource implementation enumerates these on every metrics scrape;
// the snapshot must be cheap to compute.
type PartitionGauge struct {
	Topic         string
	Partition     int32
	LeaderID      int64
	Epoch         int64
	HighWatermark int64
}

// GaugeSource snapshots the v3 runtime state that the Phase 10
// ObservableGauges sample. Implementations MUST be safe to call from a
// metrics callback (i.e. once per Prometheus scrape) and complete
// quickly — the SDK serialises callbacks per scrape, so a slow source
// stalls every other gauge.
//
// A nil GaugeSource (the default before Bootstrap wires one) makes
// every gauge report zero. That's the right answer in local-dev / test
// where there's no v3 runtime to snapshot.
type GaugeSource interface {
	// IsController returns 1 when this broker holds the cluster controller
	// lease, 0 otherwise. Sum across the fleet should always be 0 or 1
	// (during a failover briefly 0; in steady state 1).
	IsController() int64

	// AssignmentVersion is the most recent assignmentVersion this broker
	// has applied. Operators watch this for "stuck broker" detection
	// (a broker whose value stops climbing while peers advance).
	AssignmentVersion() int64

	// BrokerCountAlive is the live broker count from the broker source
	// (EndpointSlice in k8s mode). BrokerCountAssigned is the number of
	// distinct brokers in the current assignment.json. The two should
	// match in steady state; a divergence means the controller is
	// catching up.
	BrokerCountAlive() int64
	BrokerCountAssigned() int64

	// AssignmentFileSizeBytes is the on-disk size of
	// /data/__cluster/assignment.json. Crosses Kubernetes 1MB CR-status
	// cap territory at ~8000 partitions; the gauge is the
	// early-warning signal for "you're approaching the truncation bound".
	AssignmentFileSizeBytes() int64

	// Partitions returns one row per partition this broker is aware of.
	// Driven by the broker.Coordinator's snapshot — typically the union
	// of "this broker leads" and "this broker is in the assignment".
	Partitions() []PartitionGauge
}

var gaugeSource atomic.Pointer[GaugeSource]

// SetGaugeSource installs the snapshot source. Called by cmd/skafka
// after the v3 runtime is up. nil source resets to the no-op default
// (every gauge reports zero).
func SetGaugeSource(s GaugeSource) {
	if s == nil {
		gaugeSource.Store(nil)
		return
	}
	gaugeSource.Store(&s)
}

// loadGaugeSource returns the active source or nil. Used by the
// callback installed in installRuntimeGauges.
func loadGaugeSource() GaugeSource {
	p := gaugeSource.Load()
	if p == nil {
		return nil
	}
	return *p
}

// installRuntimeGauges registers the Phase 10 ObservableGauges on the
// given meter and installs a single callback that pulls every value
// from the active GaugeSource. Returns an error if any instrument
// creation fails.
//
// Called by Bootstrap once during process startup. Safe to call before
// SetGaugeSource — until a source is installed, every gauge reports 0.
func installRuntimeGauges(m metric.Meter) error {
	isController, err := m.Int64ObservableGauge("skafka.is.controller",
		metric.WithDescription("1 if this broker holds the cluster controller lease, 0 otherwise"))
	if err != nil {
		return fmt.Errorf("gauge is_controller: %w", err)
	}
	assignmentVersion, err := m.Int64ObservableGauge("skafka.assignment.version",
		metric.WithDescription("Most recent assignmentVersion applied by this broker"))
	if err != nil {
		return fmt.Errorf("gauge assignment_version: %w", err)
	}
	brokerCountAlive, err := m.Int64ObservableGauge("skafka.broker.count.alive",
		metric.WithDescription("Live brokers as observed by this broker"))
	if err != nil {
		return fmt.Errorf("gauge broker_count_alive: %w", err)
	}
	brokerCountAssigned, err := m.Int64ObservableGauge("skafka.broker.count.assigned",
		metric.WithDescription("Distinct brokers in the current assignment.json"))
	if err != nil {
		return fmt.Errorf("gauge broker_count_assigned: %w", err)
	}
	assignmentFileSize, err := m.Int64ObservableGauge("skafka.assignment.file.size",
		metric.WithDescription("Size of /data/__cluster/assignment.json"),
		metric.WithUnit("By"))
	if err != nil {
		return fmt.Errorf("gauge assignment_file_size: %w", err)
	}
	partitionLeader, err := m.Int64ObservableGauge("skafka.partition.leader",
		metric.WithDescription("Per-partition leader broker ordinal"))
	if err != nil {
		return fmt.Errorf("gauge partition_leader: %w", err)
	}
	partitionEpoch, err := m.Int64ObservableGauge("skafka.partition.epoch",
		metric.WithDescription("Per-partition leader epoch"))
	if err != nil {
		return fmt.Errorf("gauge partition_epoch: %w", err)
	}
	partitionHighWatermark, err := m.Int64ObservableGauge("skafka.partition.high.watermark",
		metric.WithDescription("Per-partition high watermark offset"))
	if err != nil {
		return fmt.Errorf("gauge partition_high_watermark: %w", err)
	}

	_, err = m.RegisterCallback(func(_ context.Context, obs metric.Observer) error {
		src := loadGaugeSource()
		if src == nil {
			// No source installed — emit zero so the gauges still appear in
			// Prometheus output (Grafana panels prefer present-but-zero over
			// missing series).
			obs.ObserveInt64(isController, 0)
			obs.ObserveInt64(assignmentVersion, 0)
			obs.ObserveInt64(brokerCountAlive, 0)
			obs.ObserveInt64(brokerCountAssigned, 0)
			obs.ObserveInt64(assignmentFileSize, 0)
			return nil
		}

		obs.ObserveInt64(isController, src.IsController())
		obs.ObserveInt64(assignmentVersion, src.AssignmentVersion())
		obs.ObserveInt64(brokerCountAlive, src.BrokerCountAlive())
		obs.ObserveInt64(brokerCountAssigned, src.BrokerCountAssigned())
		obs.ObserveInt64(assignmentFileSize, src.AssignmentFileSizeBytes())

		for _, p := range src.Partitions() {
			attrs := metric.WithAttributes(
				attribute.String("topic", p.Topic),
				attribute.Int("partition", int(p.Partition)),
			)
			obs.ObserveInt64(partitionLeader, p.LeaderID, attrs)
			obs.ObserveInt64(partitionEpoch, p.Epoch, attrs)
			obs.ObserveInt64(partitionHighWatermark, p.HighWatermark, attrs)
		}
		return nil
	},
		isController,
		assignmentVersion,
		brokerCountAlive,
		brokerCountAssigned,
		assignmentFileSize,
		partitionLeader,
		partitionEpoch,
		partitionHighWatermark,
	)
	if err != nil {
		return fmt.Errorf("register gauge callback: %w", err)
	}
	return nil
}

File: crates/kaas-observability/src/gauges.rs
//! Runtime observable gauges + [`GaugeSource`] trait.
//!
//! The
//! runtime gauges (`kaas.is.controller`, `kaas.assignment.version`,
//! per-partition leader/epoch/HWM) are registered once on the meter at
//! [`crate::bootstrap`] time; a single global [`GaugeSource`] provides
//! the snapshot at every scrape.
//!
//! A `None` source (the default before [`set_gauge_source`] is called)
//! makes every gauge report zero — dashboards prefer present-but-zero
//! over missing series.

use std::sync::Arc;

use arc_swap::ArcSwapOption;
use opentelemetry::metrics::Meter;
use opentelemetry::KeyValue;

/// One row in the per-partition gauge sample.
#[derive(Debug, Clone)]
pub struct PartitionGauge {
    pub topic: String,
    pub partition: i32,
    pub leader_id: i64,
    pub epoch: i64,
    pub high_watermark: i64,
}

/// Snapshot the v3 runtime state that the Phase 10 gauges sample.
///
/// Implementations must be safe to call from a metrics callback
/// (once per push interval) and complete quickly — the SDK serialises
/// callbacks per scrape, so a slow source stalls every other gauge.
pub trait GaugeSource: Send + Sync + 'static {
    fn is_controller(&self) -> i64;
    fn assignment_version(&self) -> i64;
    fn broker_count_alive(&self) -> i64;
    fn broker_count_assigned(&self) -> i64;
    fn assignment_file_size_bytes(&self) -> i64;
    fn partitions(&self) -> Vec<PartitionGauge>;
}

fn source_slot() -> &'static ArcSwapOption<Box<dyn GaugeSource>> {
    static SLOT: std::sync::OnceLock<ArcSwapOption<Box<dyn GaugeSource>>> =
        std::sync::OnceLock::new();
    SLOT.get_or_init(|| ArcSwapOption::from(None))
}

/// Install the snapshot source. Called by `bins/kaas::main` after
/// the v3 runtime is up. Pass `None` to reset to the no-op default
/// (every gauge reports zero).
pub fn set_gauge_source(s: Option<Box<dyn GaugeSource>>) {
    source_slot().store(s.map(Arc::new));
}

fn load_source() -> Option<Arc<Box<dyn GaugeSource>>> {
    source_slot().load_full()
}

/// Register the Phase 10 observable gauges on `meter`. Called once
/// during [`crate::bootstrap`]. Safe to call before
/// [`set_gauge_source`] — until a source is installed, every gauge
/// reports 0.
pub fn install_runtime_gauges(meter: &Meter) {
    meter
        .i64_observable_gauge("kaas.is.controller")
        .with_description("1 if this broker holds the cluster controller lease, 0 otherwise")
        .with_callback(|observer| match load_source() {
            Some(src) => observer.observe(src.is_controller(), &[]),
            None => observer.observe(0, &[]),
        })
        .build();

    meter
        .i64_observable_gauge("kaas.assignment.version")
        .with_description("Most recent assignmentVersion applied by this broker")
        .with_callback(|observer| match load_source() {
            Some(src) => observer.observe(src.assignment_version(), &[]),
            None => observer.observe(0, &[]),
        })
        .build();

    meter
        .i64_observable_gauge("kaas.broker.count.alive")
        .with_description("Live brokers as observed by this broker")
        .with_callback(|observer| match load_source() {
            Some(src) => observer.observe(src.broker_count_alive(), &[]),
            None => observer.observe(0, &[]),
        })
        .build();

    meter
        .i64_observable_gauge("kaas.broker.count.assigned")
        .with_description("Distinct brokers in the current assignment.json")
        .with_callback(|observer| match load_source() {
            Some(src) => observer.observe(src.broker_count_assigned(), &[]),
            None => observer.observe(0, &[]),
        })
        .build();

    meter
        .i64_observable_gauge("kaas.assignment.file.size")
        .with_description("Size of /data/__cluster/assignment.json")
        .with_unit("By")
        .with_callback(|observer| match load_source() {
            Some(src) => observer.observe(src.assignment_file_size_bytes(), &[]),
            None => observer.observe(0, &[]),
        })
        .build();

    meter
        .i64_observable_gauge("kaas.partition.leader")
        .with_description("Per-partition leader broker ordinal")
        .with_callback(|observer| {
            if let Some(src) = load_source() {
                for p in src.partitions() {
                    observer.observe(
                        p.leader_id,
                        &[
                            KeyValue::new("topic", p.topic),
                            KeyValue::new("partition", i64::from(p.partition)),
                        ],
                    );
                }
            }
        })
        .build();

    meter
        .i64_observable_gauge("kaas.partition.epoch")
        .with_description("Per-partition leader epoch")
        .with_callback(|observer| {
            if let Some(src) = load_source() {
                for p in src.partitions() {
                    observer.observe(
                        p.epoch,
                        &[
                            KeyValue::new("topic", p.topic),
                            KeyValue::new("partition", i64::from(p.partition)),
                        ],
                    );
                }
            }
        })
        .build();

    meter
        .i64_observable_gauge("kaas.partition.high.watermark")
        .with_description("Per-partition high watermark offset")
        .with_callback(|observer| {
            if let Some(src) = load_source() {
                for p in src.partitions() {
                    observer.observe(
                        p.high_watermark,
                        &[
                            KeyValue::new("topic", p.topic),
                            KeyValue::new("partition", i64::from(p.partition)),
                        ],
                    );
                }
            }
        })
        .build();
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used)]
mod tests {
    use super::*;

    struct FakeSource(i64);
    impl GaugeSource for FakeSource {
        fn is_controller(&self) -> i64 {
            self.0
        }
        fn assignment_version(&self) -> i64 {
            self.0 + 1
        }
        fn broker_count_alive(&self) -> i64 {
            3
        }
        fn broker_count_assigned(&self) -> i64 {
            3
        }
        fn assignment_file_size_bytes(&self) -> i64 {
            1024
        }
        fn partitions(&self) -> Vec<PartitionGauge> {
            vec![]
        }
    }

    #[test]
    fn set_and_clear_source() {
        set_gauge_source(Some(Box::new(FakeSource(1))));
        let s = load_source().unwrap();
        assert_eq!(s.is_controller(), 1);
        assert_eq!(s.assignment_version(), 2);

        set_gauge_source(None);
        assert!(load_source().is_none());
    }
}
